src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,184 @@
+
+            if (srvRecords.Count > 0)
+            {
+                var minTimeToLive = TimeSpan.FromSeconds(srvRecords.Select(s => s.InitialTimeToLive).Min());
+                if (minTimeToLive > delay)
+                {
+                    delay = minTimeToLive;
+                }
+            }
+
+            return delay;
+        }
+
+        private List<DnsEndPoint> GetDnsEndPoints(List<SrvRecord> srvRecords)
+        {
+            var endPoints = new List<DnsEndPoint>();
+
+            foreach (var srvRecord in srvRecords)
+            {
+                var host = srvRecord.Target.ToString();
+                if (host.EndsWith("."))
+                {
+                    host = host.Substring(0, host.Length - 1);
+                }
+
+                if (IsValidHost(host))
+                {
+                    // for testing
+                    if (host == "localhost.mongodb_plus_srv_test.com")
+                    {
+                        host = "localhost";
+                    }
+                    var port = srvRecord.Port;
+                    var endPoint = new DnsEndPoint(host, port);
+                    endPoints.Add(endPoint);
+                }
+            }
+
+            return endPoints;
+        }
+
+        private bool IsValidHost(string host)
+        {
+            return host.EndsWith(_parentDomainName);
+        }
+
+        private void Monitor()
+        {
+            while (true)
+            {
+                if (ShouldStopMonitoring())
+                {
+                    return;
+                }
+
+                var srvRecords = QuerySrvRecords();
+                var dnsEndPoints = GetDnsEndPoints(srvRecords);
+                _cluster.ProcessDnsResults(dnsEndPoints, _cancellationToken);
+
+                if (ShouldStopMonitoring())
+                {
+                    return;
+                }
+
+                _cancellationToken.ThrowIfCancellationRequested();
+                var delay = ComputeRescanDelay(srvRecords);
+                Thread.Sleep(delay);
+            }
+        }
+
+        private List<SrvRecord> QuerySrvRecords()
+        {
+            _cancellationToken.ThrowIfCancellationRequested();
+
+            // for testing
+            if (_lookupDomainName == "dnsmonitor_test.mongodb_plus_srv_test.com")
+            {
+                var host = DnsString.Parse("localhost.mongodb_plus_srv_test.com");
+                var info = new ResourceRecordInfo(host, ResourceRecordType.SRV, QueryClass.IN, 600, 0);
+                var srvRecord = new SrvRecord(info, 0, 0, 27017, host);
+                return new List<SrvRecord> { srvRecord };
+            }
+
+            try
+            {
+                var response = _lookupClient.Query(_query, QueryType.SRV, QueryClass.IN);
+                return response.Answers.SrvRecords().ToList();
+            }
+            catch (Exception)
+            {
+                // TODO: log exception
```

Comment:
I think we have to somehow notify the Cluster that an exception occurred to that it can be included in the ClusterDescription and, subsequently, the MongoTimeoutException thrown when server selection times out.